File: backend/src/utils/ApiResponse.js
class ApiResponse {
  constructor(statusCode, data, message = "") {
    this.statusCode = statusCode;
    this.success = statusCode < 400;
    this.message = message;
    this.data = data;
  }

  static success(data, message = "Success") {
    return new ApiResponse(200, data, message);
  }

  static created(data, message = "Created successfully") {
    return new ApiResponse(201, data, message);
  }

  static noContent(message = "Deleted successfully") {
    return new ApiResponse(204, null, message);
  }

  send(res) {
    const response = {
      success: this.success,
      message: this.message,
    };

    if (this.data !== null && this.data !== undefined) {
      response.data = this.data;
    }

    return res.status(this.statusCode).json(response);
  }
}

// Pagination response helper
class PaginatedResponse extends ApiResponse {
  constructor(data, pagination, message = "Success") {
    super(200, data, message);
    this.pagination = pagination;
  }

  static paginate(data, page, limit, total, message = "Success") {
    const totalPages = Math.ceil(total / limit);
    const pagination = {
      current_page: page,
      per_page: limit,
      total_items: total,
      total_pages: totalPages,
      has_next: page < totalPages,
      has_prev: page > 1,
    };

    const response = new PaginatedResponse(data, pagination, message);
    return response;
  }

  send(res) {
    return res.status(this.statusCode).json({
      success: this.success,
      message: this.message,
      data: this.data,
      pagination: this.pagination,
    });
  }
}

module.exports = { ApiResponse, PaginatedResponse };

File: frontend/lib/shared/models/api_response.dart
// API Response Model
class ApiResponse<T> {
  final bool success;
  final String message;
  final T? data;
  final Map<String, dynamic>? pagination;

  ApiResponse({
    required this.success,
    required this.message,
    this.data,
    this.pagination,
  });

  factory ApiResponse.fromJson(
    Map<String, dynamic> json,
    T Function(dynamic)? fromJsonT,
  ) {
    return ApiResponse(
      success: json['success'] ?? false,
      message: json['message'] ?? '',
      data: json['data'] != null && fromJsonT != null
          ? fromJsonT(json['data'])
          : json['data'],
      pagination: json['pagination'],
    );
  }
}

// Pagination Model
class Pagination {
  final int total;
  final int page;
  final int limit;
  final int totalPages;

  Pagination({
    required this.total,
    required this.page,
    required this.limit,
    required this.totalPages,
  });

  factory Pagination.fromJson(Map<String, dynamic> json) {
    return Pagination(
      total: json['total'] ?? 0,
      page: json['page'] ?? 1,
      limit: json['limit'] ?? 20,
      totalPages: json['totalPages'] ?? 1,
    );
  }

  bool get hasMore => page < totalPages;
}

// Paginated Response Model
class PaginatedResponse<T> {
  final List<T> items;
  final Pagination pagination;

  PaginatedResponse({
    required this.items,
    required this.pagination,
  });

  bool get hasMore => pagination.hasMore;
}
